## Exploring eBay Car Sales Data

In this project we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The data dictionary provided with data is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `odometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings. 

First we import pandas and numpy, then we read the csv file

In [1]:
import numpy as np
import pandas as pd
autos = pd.read_csv(r"C:\Users\youss\Desktop\DataQuest\Projects\Exploring eBay Car Sales Data\autos.csv", encoding="Latin-1")

In [2]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

After observing the dataframe info we can conclude that it has a total of 371528 entries, 20 columns, 7 with integer values, 13 with string values.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

The column names use camelcase instead of snakecase, so we will fix that using the mapping method

In [5]:
mapping_dict = {'dateCrawled':'crawled_date',
               'name':'name',
               'seller':'seller',
               'offerType':'offer_type',
               'price':'price',
               'abtest':'ab_test',
                'vehicleType':'vehicle_type',
            'yearOfRegistration':'registration_year',
               'gearbox':'gearbox',
               'powerPS':'power_ps',
                'model':'model',
                'kilometer':'kilometers',
            'monthOfRegistration':'registration_month',
             'fuelType':'fuel_type',
             'brand':'brand',
             'notRepairedDamage':'unrepaired_damage',
              'dateCreated':'ad_created',
              'nrOfPictures':'pictures_num',
               'postalCode':'postal_code',
                'lastSeen':'last_seen'}
autos.columns = autos.columns.map(mapping_dict)

In [6]:
autos.head()

crawled_date                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  registration_year    gearbox  power_ps  model  \
0    480    test          NaN               1993    manuell         0   golf   
1  18300    test        coupe               2011    manuell       190    NaN   
2   9800    test          suv               2004  automatik       163  grand   
3   1500    test   kleinwagen               2001    manuell        75   golf   
4   3600    test   kleinwagen               2008    manuell        69  fabia   

   kilometers  registration_month fuel_type       brand unrepaired_damage  \
0      150000                   0    benzin  volkswagen               NaN   
1      125000                   5    diesel        audi                ja   
2      125000                   8    diesel        jeep               NaN   
3      150000                   6    benzin  volkswagen              nein   
4       90000                   7    diesel       skoda              nein   

            ad_created  pictures_num  postal_code            last_seen  
0  2016-03-24 00:00:00             0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0        60437  2016-04-06 10:17:21

Lets's investigate the data

In [7]:
autos.describe(include='all')

crawled_date         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model     kilometers  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  pictures_num   postal_code  \
count             299468               371528      371528.0  371528.00000   
unique                 2                  114           NaN           NaN   
top                 nein  2016-04-03 00:00:00           NaN           NaN   
freq              263182                14450           NaN           NaN   
mean                 NaN                  NaN           0.0   50820.66764   
std                  NaN                  NaN           0.0   25799.08247   
min                  NaN                  NaN           0.0    1067.00000   
25%                  NaN                  NaN           0.0   30459.00000   
50%                  NaN                  NaN           0.0   49610.00000   
75%                  NaN                  NaN           0.0   71546.00000   
max                  NaN                  NaN           0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [8]:
autos.dtypes

crawled_date          object
name                  object
seller                object
offer_type            object
price                  int64
ab_test               object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
kilometers             int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
pictures_num           int64
postal_code            int64
last_seen             object
dtype: object

In [9]:
autos['pictures_num'].value_counts()

pictures_num
0    371528
Name: count, dtype: int64

In [10]:
autos['seller'].value_counts()

seller
privat        371525
gewerblich         3
Name: count, dtype: int64

After investigation we see that most values are written in german not english so that needs cleaning, also all rows in the `pictures_num` column has all values of 0 so we will remove the whole column and `seller` column will be removed too as almost the whole column has the same value.

In [11]:
# removing non useful columns
autos = autos.drop('pictures_num', axis=1)
autos = autos.drop('seller', axis=1)

In [12]:
autos.dtypes

crawled_date          object
name                  object
offer_type            object
price                  int64
ab_test               object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
kilometers             int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
postal_code            int64
last_seen             object
dtype: object

In [13]:
autos['offer_type'].value_counts()

offer_type
Angebot    371516
Gesuch         12
Name: count, dtype: int64

In [14]:
autos['vehicle_type'].value_counts()

vehicle_type
limousine     95894
kleinwagen    80023
kombi         67564
bus           30201
cabrio        22898
coupe         19015
suv           14707
andere         3357
Name: count, dtype: int64

In [15]:
autos['gearbox'].value_counts()

gearbox
manuell      274214
automatik     77105
Name: count, dtype: int64

In [16]:
autos['fuel_type'].value_counts()

fuel_type
benzin     223857
diesel     107746
lpg          5378
cng           571
hybrid        278
andere        208
elektro       104
Name: count, dtype: int64

In [17]:
autos['unrepaired_damage'].value_counts()

unrepaired_damage
nein    263182
ja       36286
Name: count, dtype: int64

In [18]:
cleaning_dict1 = {'Angebot':'offer', 'Gesuch':'request'}

cleaning_dict2 = {'limousine':'limousine',
                'kleinwagen':'subcompact',
                'kombi':'van',
                'bus':'bus',
                'cabrio':'convertible',
                'coupe':'coupe',
                'suv':'suv',
                'andere':'others'}

cleaning_dict3 = {'manuell':'manual', 'automatik':'automatic'}

cleaning_dict4 = {'benzin':'petrol',
                'diesel':'diesel',
                'lpg':'lpg',
                'cng':'cng',
                'hybrid':'hybrid',
                'andere':'others',
                'elektro':'electricity'}

cleaning_dict5 = {'nein':'no', 'ja':'yes'}

autos['offer_type'] = autos['offer_type'].map(cleaning_dict1)
autos['vehicle_type'] = autos['vehicle_type'].map(cleaning_dict2)
autos['gearbox'] = autos['gearbox'].map(cleaning_dict3)
autos['fuel_type'] = autos['fuel_type'].map(cleaning_dict4)
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(cleaning_dict5)

In [19]:
autos

crawled_date                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

       offer_type  price  ab_test vehicle_type  registration_year    gearbox  \
0           offer    480     test          NaN               1993     manual   
1           offer  18300     test        coupe               2011     manual   
2           offer   9800     test          suv               2004  automatic   
3           offer   1500     test   subcompact               2001     manual   
4           offer   3600     test   subcompact               2008     manual   
...           ...    ...      ...          ...                ...        ...   
371523      offer   2200     test          NaN               2005        NaN   
371524      offer   1199     test  convertible               2000  automatic   
371525      offer   9200     test          bus               1996     manual   
371526      offer   3400     test          van               2002     manual   
371527      offer  28990  control    limousine               2013     manual   

        power_ps        model  kilometers  registration_month fuel_type  \
0              0         golf      150000                   0    petrol   
1            190          NaN      125000                   5    diesel   
2            163        grand      125000                   8    diesel   
3             75         golf      150000                   6    petrol   
4             69        fabia       90000                   7    diesel   
...          ...          ...         ...                 ...       ...   
371523         0          NaN       20000                   1       NaN   
371524       101       fortwo      125000                   3    petrol   
371525       102  transporter      150000                   3    diesel   
371526       100         golf      150000                   6    diesel   
371527       320      m_reihe       50000                   8    petrol   

                 brand unrepaired_damage           ad_created  postal_code  \
0           volkswagen               NaN  2016-03-24 00:00:00        70435   
1                 audi               yes  2016-03-24 00:00:00        66954   
2                 jeep               NaN  2016-03-14 00:00:00        90480   
3           volkswagen                no  2016-03-17 00:00:00        91074   
4                skoda                no  2016-03-31 00:00:00        60437   
...                ...               ...                  ...          ...   
371523  sonstige_autos               NaN  2016-03-14 00:00:00        39576   
371524           smart                no  2016-03-05 00:00:00        26135   
371525      volkswagen                no  2016-03-19 00:00:00        87439   
371526      volkswagen               NaN  2016-03-20 00:00:00        40764   
371527             bmw                no  2016-03-07 00:00:00        73326   

                  last_seen  
0       2016-04-07 03:16:57  
1       2016-04-07 01:46:50  
2       2016-04-05 12:47:46  
3       2016-03-17 17:40:17  
4       2016-04-06 10:17:21  
...                     ...  
371523  2016-04-06 00:46:52  
371524  2016-03-11 18:17:12  
371525  2016-04-07 07

Now let's see if there is any outliers in the numeric values.

In [20]:
autos['price'].max()

2147483647

In [21]:
print(autos['power_ps'].max())
print(autos['power_ps'].min())

20000
0


Here we can see that the `price` and `power_ps` columns have outliers so we need to clean that.

In [22]:
autos['price'].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [23]:
autos = autos[autos['price'] <= 1500000].dropna()

After observing some data about the `price` column, I decided that any row with price higher than 1.5 million should be removed.

In [24]:
autos['power_ps'].describe()

count    260946.000000
mean        126.330317
std         145.278946
min           0.000000
25%          78.000000
50%         116.000000
75%         150.000000
max       20000.000000
Name: power_ps, dtype: float64

In [25]:
autos = autos[(autos['power_ps'] >= 50) & (autos['power_ps'] <= 500)].dropna()

After observing some data from `power_ps` column, I decided to remove all rows that have values less than 50 and more than 500 in the `power_ps` column.

In [26]:
print(autos['crawled_date'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

crawled_date
2016-03-05    0.026446
2016-03-06    0.014699
2016-03-07    0.035918
2016-03-08    0.032971
2016-03-09    0.033630
2016-03-10    0.032106
2016-03-11    0.032623
2016-03-12    0.036868
2016-03-13    0.016259
2016-03-14    0.036003
2016-03-15    0.032599
2016-03-16    0.029563
2016-03-17    0.030990
2016-03-18    0.012685
2016-03-19    0.034871
2016-03-20    0.036399
2016-03-21    0.035352
2016-03-22    0.031839
2016-03-23    0.032174
2016-03-24    0.029951
2016-03-25    0.032587
2016-03-26    0.032389
2016-03-27    0.030222
2016-03-28    0.035227
2016-03-29    0.034632
2016-03-30    0.033254
2016-03-31    0.031952
2016-04-01    0.034746
2016-04-02    0.035461
2016-04-03    0.039605
2016-04-04    0.038618
2016-04-05    0.012423
2016-04-06    0.003311
2016-04-07    0.001629
Name: proportion, dtype: float64


In [27]:
print(autos['crawled_date'].str[:10].value_counts(normalize=True, dropna=False).sort_index().describe())

count    34.000000
mean      0.029412
std       0.009700
min       0.001629
25%       0.030019
50%       0.032593
75%       0.035138
max       0.039605
Name: proportion, dtype: float64


After observing some statistical data for the `crawled_date` column we can conclude that:
- The majority of the values (75%) fall within the range of 0.030018 to 0.035124, suggesting a relatively narrow spread for most of the data.
- The dataset is right-skewed, as the mean is less than the median, and the maximum value is farther from the mean than the minimum value.

In [28]:
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

ad_created
2015-03-20    0.000004
2015-08-07    0.000004
2015-08-10    0.000004
2015-09-04    0.000008
2015-09-09    0.000004
                ...   
2016-04-03    0.039827
2016-04-04    0.038634
2016-04-05    0.011291
2016-04-06    0.003295
2016-04-07    0.001577
Name: proportion, Length: 105, dtype: float64


In [29]:
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index().describe())

count    105.000000
mean       0.009524
std        0.014763
min        0.000004
25%        0.000008
50%        0.000028
75%        0.029870
max        0.039827
Name: proportion, dtype: float64


After observing some statistical data for the `ad_created` column we can conclude that:
- The majority of the values (75%) fall within the range of 0.000008 to 0.029860, suggesting a wide spread of values in the dataset.
- The data is right-skewed, as the mean is less than the median, and the maximum value is farther from the mean than the minimum value.
- The range between the 25th and 75th percentiles is substantial, indicating variability in the central part of the distribution.

In [30]:
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

last_seen
2016-03-05    0.001193
2016-03-06    0.003877
2016-03-07    0.004912
2016-03-08    0.007244
2016-03-09    0.008853
2016-03-10    0.010361
2016-03-11    0.012071
2016-03-12    0.022141
2016-03-13    0.007964
2016-03-14    0.011566
2016-03-15    0.015406
2016-03-16    0.015479
2016-03-17    0.027190
2016-03-18    0.006614
2016-03-19    0.015248
2016-03-20    0.018773
2016-03-21    0.019190
2016-03-22    0.019432
2016-03-23    0.017047
2016-03-24    0.018414
2016-03-25    0.017880
2016-03-26    0.014913
2016-03-27    0.015657
2016-03-28    0.021062
2016-03-29    0.022626
2016-03-30    0.022905
2016-03-31    0.023552
2016-04-01    0.023443
2016-04-02    0.023984
2016-04-03    0.024684
2016-04-04    0.024938
2016-04-05    0.132866
2016-04-06    0.230739
2016-04-07    0.137777
Name: proportion, dtype: float64


In [31]:
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index().describe())

count    34.000000
mean      0.029412
std       0.046048
min       0.001193
25%       0.011692
50%       0.018147
75%       0.023308
max       0.230739
Name: proportion, dtype: float64


After observing some statistical data for the `last_seen` column we can conclude that:
- The majority of the values (75%) fall within the range of 0.011690 to 0.023310, indicating a relatively narrower spread for most of the data.
- The data is right-skewed, as the mean is less than the median, and the maximum value is farther from the mean than the minimum value.
- The maximum value of 0.230788 is significantly higher than the mean and median, indicating a potential outlier or a distinct event where items were last seen.

Now let's change the values in the `crawled_date`, `ad_created`, and `last_seen` columns from string to integer.

The format will be as follows Year/Month/Day ex:
- 2016-03-17 will be 20160317

In [32]:
autos['crawled_date'] = autos['crawled_date'].str[:10].replace("")
autos['crawled_date'] = autos['crawled_date'].str.replace('-', '')
autos['crawled_date'] = autos['crawled_date'].astype(int)
autos['crawled_date']

3         20160317
4         20160331
5         20160404
6         20160401
7         20160321
            ...   
371517    20160328
371520    20160319
371524    20160305
371525    20160319
371527    20160307
Name: crawled_date, Length: 247370, dtype: int32

In [33]:
autos['ad_created'] = autos['ad_created'].str[:10].replace("")
autos['ad_created'] = autos['ad_created'].str.replace('-', '')
autos['ad_created'] = autos['ad_created'].astype(int)
autos['ad_created']

3         20160317
4         20160331
5         20160404
6         20160401
7         20160321
            ...   
371517    20160328
371520    20160319
371524    20160305
371525    20160319
371527    20160307
Name: ad_created, Length: 247370, dtype: int32

In [34]:
autos['last_seen'] = autos['last_seen'].str[:10].replace("")
autos['last_seen'] = autos['last_seen'].str.replace('-', '')
autos['last_seen'] = autos['last_seen'].astype(int)
autos['last_seen']

3         20160317
4         20160406
5         20160406
6         20160405
7         20160325
            ...   
371517    20160402
371520    20160319
371524    20160311
371525    20160407
371527    20160322
Name: last_seen, Length: 247370, dtype: int32

Based on all the information I gained, I conclude that no car can be registered after it has been seen, the `last_seen` column has a maximum year of 2016, so no car can be registered after 2016, that's why I decided to remove any rows that has a value more than 2016 in the `registration_year` column.

In [35]:
print(autos['registration_year'].max())
print(autos['registration_year'].min())

2018
1930


In [36]:
autos = autos[autos['registration_year'] < 2016].dropna()

Now I am going to extract the brand names from the `name` column which is usually the first string value and save it to a new column named `brands`

In [37]:
autos['brands'] = (autos['name'].str.split().str[0])
autos['brands'] = autos['brands'].str.replace('_', " ")
autos['brands'] = autos['brands'].str.split().str[0]
autos['brands']

3               GOLF
4              Skoda
5                BMW
6            Peugeot
7                 VW
             ...    
371517    Volkswagen
371520         turbo
371524         Smart
371525    Volkswagen
371527           BMW
Name: brands, Length: 247048, dtype: object

In [38]:
autos['brands'] = autos['brands'].replace('VW', 'Volkswagen')
autos['brands'] = autos['brands'].replace('Golf', 'Volkswagen')
autos['brands'] = autos['brands'].replace('GOLF', 'Volkswagen')

Let's check if there are any null values in the columns.

In [39]:
autos.isnull().sum()

crawled_date          0
name                  0
offer_type            0
price                 0
ab_test               0
vehicle_type          0
registration_year     0
gearbox               0
power_ps              0
model                 0
kilometers            0
registration_month    0
fuel_type             0
brand                 0
unrepaired_damage     0
ad_created            0
postal_code           0
last_seen             0
brands                0
dtype: int64

Our columns are clear of non values and our data cleaning process is finished, now let's save the cleaned dataframe into a new csv for further analysis and summarize the data cleaning process.

In [40]:
autos.to_csv('cleaned_autos.csv', index=False)

**Data Cleaning Summary**
- Changed the column names from camelcase to snakecase.
- Removed columns that have almost the same exact value in all rows.
- Translated German values to English.
- Removing outliers from `price` and `power_ps` columns.
- Changed the values in the `crawled_date`, `ad_created`, and `last_seen` columns from string to integer.
- Removed all rows with a `registration_year` more than 2016.
- Extracted the brand names from the `name` column and saved it to a new column named `brands`.
- Corrected some in accurate names in the `brands` column.

**Exploring price by brand**

In [41]:
brands_counts = autos['brands'].value_counts()
print(brands_counts)

brands
Volkswagen         45407
BMW                26134
Mercedes           22857
Audi               22262
Opel               21143
                   ...  
Tausch/verkaufe        1
TOLLEN                 1
Suessen                1
civic                  1
turbo                  1
Name: count, Length: 3444, dtype: int64


In [42]:
top_brands = brands_counts.head(15).index
print(top_brands)

Index(['Volkswagen', 'BMW', 'Mercedes', 'Audi', 'Opel', 'Ford', 'Renault',
       'Peugeot', 'Fiat', 'Seat', 'Skoda', 'Mazda', 'Toyota', 'Nissan',
       'Hyundai'],
      dtype='object', name='brands')


In [43]:
aggregate_data = {}

for brand in top_brands:
    brand_subset = autos[autos['brands'] == brand]
    mean_price = brand_subset['price'].mean()
    aggregate_data[brand] = mean_price

print("Dictionary of Aggregate Data:")
print(aggregate_data)

Dictionary of Aggregate Data:
{'Volkswagen': 6639.991388992887, 'BMW': 9734.849812504783, 'Mercedes': 9509.565472284201, 'Audi': 10353.109783487558, 'Opel': 3710.8284065648204, 'Ford': 4565.643902105551, 'Renault': 3073.5823680823682, 'Peugeot': 3866.912906610703, 'Fiat': 3527.310889929742, 'Seat': 5326.49678751721, 'Skoda': 7187.623603691112, 'Mazda': 4894.16572077185, 'Toyota': 5734.276511698572, 'Nissan': 5629.988711819389, 'Hyundai': 6304.670463791701}


Analysis:
- The dictionary provides the mean prices for the top 15 brands in the given DataFrame. This information can be valuable for understanding the average pricing trends for each brand. Brands with higher mean prices may indicate premium products, while lower mean prices may suggest more affordable options. The analysis can be further extended by considering additional factors such as sales volume, customer ratings, or product categories to gain a comprehensive understanding of the market dynamics.